packages/convex-helpers/server/zod4.ts

```
@@ -0,0 +1,642 @@
+const registryHelpers = {
+  getMetadata: (schema: z.ZodTypeAny) => _meta.get(schema),
+  setMetadata: (schema: z.ZodTypeAny, meta: { isConvexId?: boolean; tableName?: string }) =>
+    _meta.set(schema, meta),
+};
+
+export function zid<
+  DataModel extends GenericDataModel,
+  TableName extends TableNamesInDataModel<DataModel> = TableNamesInDataModel<DataModel>,
+>(tableName: TableName) {
+  const base = z
+    .string()
+    .refine((s) => typeof s === "string" && s.length > 0, {
+      message: `Invalid ID for table "${tableName}"`,
+    })
+    .transform((s) => s as GenericId<TableName>)
+    .brand(`ConvexId_${tableName}`)
+    .describe(`convexId:${tableName}`);
+  registryHelpers.setMetadata(base, { isConvexId: true, tableName });
+  return base as z.ZodType<GenericId<TableName>>;
+}
+
+export const withSystemFields = <
+  Table extends string,
+  T extends { [key: string]: z.ZodTypeAny },
+>(tableName: Table, zObject: T) => {
+  return { ...zObject, _id: zid(tableName as any), _creationTime: z.number() } as const;
+};
+
+function isZid(schema: z.ZodTypeAny): boolean {
+  const m = registryHelpers.getMetadata(schema);
+  return !!(m && m.isConvexId && typeof m.tableName === "string");
+}
+
+function makeUnion(members: GenericValidator[]) {
+  const arr = members.filter((m): m is GenericValidator => !!m);
+  if (arr.length === 0) return v.any();
+  if (arr.length === 1) return arr[0];
+  const [a, b, ...rest] = arr as [GenericValidator, GenericValidator, ...GenericValidator[]];
+  return v.union(a, b, ...rest);
+}
+
+// Narrow helpers for Convex validators
+function asValidator(x: unknown): any {
+  return x as unknown as any;
+}
+
+// Zod v4 -> Convex (input)
+function zodToConvexInternal(
+  zodValidator: z.ZodTypeAny,
+  visited: Set<z.ZodTypeAny> = new Set(),
+): GenericValidator {
+  if (!zodValidator) return v.any();
+  if (visited.has(zodValidator)) return v.any();
```

What is the goal of this?

It’s true that Zod now supports [recursive objects](https://zod.dev/v4?id=recursive-objects). However, there is no equivalent thing with Convex validators at the moment. I think that we should throw an error message instead of this happens.